### PR TITLE
Remote gallery viewer - a library for painting

### PR DESCRIPTION
**Pull request by ShiftyRail**
## What this does

This adds a computer, similar to the existing library database, to store, browse and retrieve paintings done during previous rounds. Somewhat unatomically, I've also added a "description" field to the library just to that you can add a short summary to books you are uploading. Unfortunately, all existing books will have a *blank* description as a consequence. I deemed it necessary for using the OOP methods of the library computer, which I made ever so slightly more modular.

Images:

image

image

One downside is that I add to bundle it up onto a new computer instead of using the existing checkout computer. Although I /believe/ that with just a little bit of effort I can merge this new computer with the existing library database. I've not made up my mind yet about it. Here's what it looks like when mapped on Box and Meta.

image

image

## Why it's good

Your paintings will now live on forever (almost) and you can retrieve previous ones for decoration. Gives the librarian more stuff to do. Paintings are popular.

Depending on feedback, I will consider merging this stuff with the existing computer and revert the mapping changes. 

Additionally, I will (but in another PR) consider adding an upvote/downvote (or just plain upvote, once per ckey) system to both the library and the painting archive. This is *relatively* easy to do in SQL.

## Changelog
:cl:
 * rscadd: Added remote gallery browsers in the Library! You can now upload and order the finest work of art in all the galaxy and proudly display it to decorate your brig.
 * tweak: Small mapping changes to accommodate for the new library computer.

### CODE REVIEW

File: code/modules/html_interface/paintTool/custom_painting_datum.dm [hunk 1]
```diff
@@ -264,6 +265,31 @@
...
+		painting.bitmap
+	)
+	return json_encode(L)
+
+/proc/json2painting(var/json_data, var/title, var/author, var/description)
+	var/list/L = json_decode(json_data)
+	var/datum/custom_painting/painting = new(null, L[1], L[2], L[3], L[4], L[5]) // no parents
+	var/list/bitmap_to_copy = L[6]
+	painting.bitmap = bitmap_to_copy.Copy()
+	painting.title = title
```
Comment (SonixApache) [inline code/modules/html_interface/paintTool/custom_painting_datum.dm:287]:
```suggestion
	painting.title = title + " (Reprint)"
```

**Comment by jwhitak** [comment]
I'd love for the two to be merged, but I understand if they can't or if it'd take too much effort.

**Comment by ShiftyRail** [comment]
Open to ideas: what should the (inevitable) emag function of this be?

**Comment by Blithering** [comment]
Emagged painting library? Vox feet pics.

**Comment by PrimeDSS13** [comment]
I'd merge them together, but if it's not possible at least get a different sprite for the art terminal. It just looks silly having two identical computers on the desk.
For the emagged computer, maybe it just prints out a random painting? Alternatively if you want something that takes more work to code, maybe it creates a painting with randomized pixel colors?

**Comment by SonixApache** [comment]
emags don't have to do epic magical shit
if there's no lock on this, there doesn't need to be an emag act

**Comment by Eneocho** [comment]
Do you retrieve the painting pattern so you can paint it yourself, retrieve the painting as a black and white printing or retrieve the painting as an identical copy of the original? If the last one, kino. If the first one, that's alright, if the black and white one, please no.

**Comment by Inorien** [comment]
for emag, would it be too difficult to have it randomly mix different paintings?
or just introduce a ton of corruptions to the print 
if only to antagonise the person who wanted to print out a painting

File: __DEFINES/global.dm [hunk 2]
```diff
@@ -179,13 +179,13 @@ var/datum/nanomanager/nanomanager = new()
 
 var/sqladdress = "localhost"
 var/sqlport = 3306
-var/sqldb = "tgstation"
+var/sqldb = "feedback"
```
Comment (ShiftyRail) [inline __DEFINES/global.dm:182]: To fix before merged

**Comment by RealestEstate** [comment]
art computer
its a lazy recolor but it should help visually distinguish the two library computers from each other.
what do you think?

**Comment by ShiftyRail** [comment]
Looks a lot better than what I could've done, tbh, but I would maybe make it a bit less blocky and more sleek.

**Comment by RealestEstate** [comment]
sure, I'll see what I can do

**Comment by NigglyWiggly** [comment]
>Adding a second identical looking computer on top of the Librarian's desk

Why not just replace the rental computer around the corner that nobody ever uses?

File: code/modules/library/computers/remote_gallery_access.dm [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	return scanner.cached_painting
+
+/obj/machinery/computer/library/checkout/remote_gallery/make_external_book(var/datum/cachedbook/newbook)
+	message_admins("printing [newbook.title]...")
+	if(!newbook)
+		return
+	var/obj/item/mounted/frame/painting/custom/C = new(get_turf(src))
+	C.name = "[newbook.title] by [newbook.author]"
+	C.desc = newbook.description
+	C.painting_data = json2painting(newbook.content, newbook.title, newbook.author, newbook.description)
```
Comment (jellyveggie2) [inline code/modules/library/computers/remote_gallery_access.dm]:
```suggestion
	C.set_painting_data(json2painting(newbook.content, newbook.title, newbook.author, newbook.description))
```
Setting the data through `set_painting_data()` will ensure `C.painting_data.set_parent(C)` gets called and the multipart handler gets set. 
...Don't remember what exactly would happen if it's not set, probably just being unable to paint until a player hangs the painting (likely the first thing most players will do).
So nothing serious, but still.

**Comment by ShiftyRail** [comment]
@Kurfursten This is all ready for merging + it has your art vendor mapped in as well

**Comment by NigglyWiggly** [comment]
You left it on the fucking desk, how dare you

**Pull request merged**